murmur.go

```diff
@@ -30,6 +35,24 @@ type digest struct {
 
 func (d *digest) BlockSize() int { return 1 }
 
+func (d *digest) WriteString(s string) (int, error) {
+	// This code does the same as:
+	//
+	//   return d.Write([]byte(s))
+	//
+	// because the parameter `p` to Write is passed to d.bmix, which is an
+	// interface type, the simple conversion to a byte slice forces the compiler
+	// to make a heap allocation and copy the slice. The use of unsafe here lets
+	// us take over the default behavior of the compiler to have the byte slice
+	// share the underlying memory buffer of the string and avoid the extra heap
+	// allocation.
+	return d.Write(*(*[]byte)(unsafe.Pointer(&reflect.SliceHeader{
```

Comment:
This conversion is unsafe (see golang/go#40701)

Reply:
I believe this case is correct because it is covered by the rules 3 and 6 of unsafe:

> (3) Conversion of a Pointer to a uintptr and back, with arithmetic.
>
> If p points into an allocated object, it can be advanced through the object by conversion to uintptr, addition of an offset, and conversion back to Pointer.

> (6) Conversion of a reflect.SliceHeader or reflect.StringHeader Data field to or from Pointer.
>
> As in the previous case, the reflect data structures SliceHeader and StringHeader declare the field Data as a uintptr to keep callers from changing the result to an arbitrary type without first importing "unsafe". However, this means that SliceHeader and StringHeader are only valid when interpreting the content of an actual slice or string value.

https://golang.org/pkg/unsafe/#Pointer

Let me know if I'm missing something.

Reply:
It's hard for me to actually tell and I wish this type of conversion were documented as explicitly safe or unsafe. I took this as violating rule 6 because the slice doesn't refer to an actual underlying slice, the slice is being made up out of thin air from a reflect.SliceHeader. What's more obviously safe to me is when the slice header is referring to an underlying slice beforehand.

It is entirely possible that this is safe, I just haven't trusted it since the six unsafe rules came out (this was my original go to for how to convert strings to slices).

As an aside, I think the 40701 I mentioned in the first ticket refers to other behavior that this is definitely _not_ doing: taking in a byte pointer and making a slice out of that.

Reply:
https://github.com/jlauinger/go-safer

I believe this is effectively patterns 1 and 2.

Reply:
I don't think it is pattern 1 and 2, because the conversion between to and from unsafe is done in a single expression.

Reply:
Actually, looks like the Go doc may have been updated to explicitly indicate that this conversion may not be safe:

> In general, reflect.SliceHeader and reflect.StringHeader should be used only as *reflect.SliceHeader and *reflect.StringHeader pointing at actual slices or strings, never as plain structs. A program should not declare or allocate variables of these struct types.

Here's an alternative approach we could try:
https://github.com/cespare/xxhash/blob/master/xxhash_unsafe.go#L52-L57

```go
// sliceHeader is similar to reflect.SliceHeader, but it assumes that the layout
// of the first two words is the same as the layout of a string.
type sliceHeader struct {
	s   string
	cap int
}
```

What do you think?

Reply:
I think something like https://github.com/twmb/murmur3/commit/610077ff6fd864908e4604f64b6c0fc000aa6232 is safer, and afaict it's just as fast.